nengo/_vendor/npconv2d/conv2d.py

```diff
@@ -0,0 +1,132 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import numpy as np
+
+
+def calc_pad(pad, in_siz, out_siz, stride, ksize):
+    """Calculate padding width.
+    Args:
+        pad: padding method, "SAME", "VALID", or manually speicified.
+        ksize: kernel size [I, J].
+    Returns:
+        pad_: Actual padding width.
+    """
+    if pad == 'SAME':
+        return (out_siz - 1) * stride + ksize - in_siz
+    elif pad == 'VALID':
+        return 0
+    else:
+        return pad
+
+
+def calc_size(h, kh, pad, sh):
+    """Calculate output image size on one dimension.
+    Args:
+        h: input image size.
+        kh: kernel size.
+        pad: padding strategy.
+        sh: stride.
+    Returns:
+        s: output size.
+    """
+
+    if pad == 'VALID':
+        return np.ceil((h - kh + 1) / sh)
+    elif pad == 'SAME':
+        return np.ceil(h / sh)
+    else:
+        return int(np.ceil((h - kh + pad + 1) / sh))
+
+
+def extract_sliding_windows(x, ksize, pad, stride, floor_first=True):
+    """Converts a tensor to sliding windows.
+    Args:
+        x: [N, H, W, C]
+        k: [KH, KW]
+        pad: [PH, PW]
+        stride: [SH, SW]
+    Returns:
+        y: [N, (H-KH+PH+1)/SH, (W-KW+PW+1)/SW, KH * KW, C]
+    """
+    n = x.shape[0]
+    h = x.shape[1]
+    w = x.shape[2]
+    c = x.shape[3]
+    kh = ksize[0]
+    kw = ksize[1]
+    sh = stride[0]
+    sw = stride[1]
+
+    h2 = int(calc_size(h, kh, pad, sh))
+    w2 = int(calc_size(w, kw, pad, sw))
+    ph = int(calc_pad(pad, h, h2, sh, kh))
+    pw = int(calc_pad(pad, w, w2, sw, kw))
+
+    ph0 = int(np.floor(ph / 2))
+    ph1 = int(np.ceil(ph / 2))
+    pw0 = int(np.floor(pw / 2))
+    pw1 = int(np.ceil(pw / 2))
+
+    if floor_first:
+        pph = (ph0, ph1)
+        ppw = (pw0, pw1)
+    else:
+        pph = (ph1, ph0)
+        ppw = (pw1, pw0)
+    x = np.pad(
+        x, ((0, 0), pph, ppw, (0, 0)),
+        mode='constant',
+        constant_values=(0.0,))
+
+    y = np.zeros([n, h2, w2, kh, kw, c])
+    for ii in range(h2):
+        for jj in range(w2):
+            xx = ii * sh
+            yy = jj * sw
+            y[:, ii, jj, :, :, :] = x[:, xx:xx + kh, yy:yy + kw, :]
+    return y
+
+
+def conv2d(x, w, pad='SAME', stride=(1, 1)):
+    """2D convolution (technically speaking, correlation).
+    Args:
+        x: [N, H, W, C]
+        w: [I, J, C, K]
+        pad: [PH, PW]
+        stride: [SH, SW]
+    Returns:
+        y: [N, H', W', K]
+    """
+    ksize = w.shape[:2]
+    x = extract_sliding_windows(x, ksize, pad, stride)
+    ws = w.shape
+    w = w.reshape([ws[0] * ws[1] * ws[2], ws[3]])
+    xs = x.shape
+    x = x.reshape([xs[0] * xs[1] * xs[2], -1])
+    y = x.dot(w)
+    y = y.reshape([xs[0], xs[1], xs[2], -1])
+    return y
```

Comment:
I'm not convinced this is the best way to do convolution. Did you test it for speed and memory use? It looks to me like `extract_sliding_windows` will create a version of `x` that is potentially quite large. Its shape is `(n, h2, w2, kh, kw, c)`, where h2 and w2 are the output size (fine), but kh and kw are the size of the kernel. If you've got e.g. a 9x9 kernel, then this intermediary object is going to use 81 times as much memory as the output object (not accounting for any difference in input and output channels). The advantage to this method is that you can line everything up to run it through a single `dot` call, which may save time. I'll try to do some profiling on this to see how much potential time saving there is.

Reply:
It does definitely save time, but I haven't done any memory profiling to see how it compares to other options in that regard.

Reply:
I added a commit that uses a view instead of copying in `extract_sliding_windows`, so the memory should no longer be an issue. And as per your profiling, the speed is much faster.